### PR TITLE
insert check for one line sample sheet

### LINKED ISSUE #589: [BUG] Sarek-dev skips directly to multiqc after interval split
## Check Documentation

I have checked the following places for your error:

- [x ] [nf-core website: troubleshooting](https://nf-co.re/usage/troubleshooting)
- [x ] [nf-core/sarek pipeline documentation](https://nf-co.re/sarek/usage)

## Description of the bug

When running the current sarek dev branch with a csv formatted like this 
```
HG0001,XX,0,HG0001,1,/home/ubuntu/U0a_CGATGT_L002/U0a_CGATGT_L002_R1_001.fastq.gz,/home/ubuntu/U0a_CGATGT_L002/U0a_CGATGT_L002_R2_001.fastq.gz 
```
 the process starts fine, but then after `NFCORE_SAREK:sarek:PREPARE_INTERVALS:TABIX_BGZIPTABIX_INTERVAL_SPLIT` it jumps straight to multiQC and makes no useful output. Further discussion in [this](https://nfcore.slack.com/archives/C02MDBZAYJK/p1655121583870879) thread and [this](https://nfcore.slack.com/archives/C02MDBZAYJK/p1655193790239759).

It's probably because the header line is missing

## Steps to reproduce

```
#Minimal reproducible example

#1 start aws c5.24xlarge 96-core machine using ECS-optimized AWS image (has docker etc)

#2 get input data
Download U0a_CGATGT_L002_R1_001.fastq.gz (and R2) from GiAB web-site


#3 get miniconda
curl -O https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh
sh Miniconda3-latest-Linux-x86_64.sh


#4 install nextflow (22.04.3) using curl instructions from www.nextflow.com
curl -fsSL get.nextflow.io | bash
mkdir bin
mv nextflow ~/bin/



#5 get Sarek 
git clone git@github.com:nf-core/sarek.git 
cd sarek
git checkout dev (hash 7b304c3f9a6e4257bb65176155740b66ea85ec81)
git checkout 4420a02d15be691f54cf6d5ef68dcf4acc0d6ba8

#6 make a samples.csv that looks like this
HG0001,XX,0,HG0001,1,/home/ubuntu/U0a_CGATGT_L002/U0a_CGATGT_L002_R1_001.fastq.gz,/home/ubuntu/U0a_CGATGT_L002/U0a_CGATGT_L002_R2_001.fastq.gz 


#7 run 
nextflow run main.nf \
-profile docker -work-dir ~/data/work/‘ \
-—tools strelka \
—vep_cache_version 104
```
## Expected behaviour

Don't silently skip all the alignment and calling steps.

## Log files

Have you provided the following extra information/files:

- [x] The command used to run the pipeline
- [ ] The `.nextflow.log` file 

## System

- Hardware: c5.24xlarge 96-core
- Executor: tested with local and awsbatch
- OS: Ubuntu Linux AWS 
- Version: 22.04 LTS

## Nextflow Installation

- Version: (22.04.3)

## Container engine

- Engine: docker
- version:  20.10.12

## Additional context

It would be nice if a user-friendly error message explaining what to do.

### PR DESCRIPTION
**Pull request by lassefolkersen**
This fixes #589 by checking that sample sheet is at least two lines. If it's only 1 line, the subsequent `splitCsv(header: true)` call will read it as empty and skip all other samplesheet checks. Also - upgraded missing `patient` and `sample` cases from log.warn to log.error.

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
    - [ ] If you've added a new tool - add to the software_versions process and a regex to `scrape_software_versions.py`
    - [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/sarek/tree/master/.github/CONTRIBUTING.md)
    - [ ] If necessary, also make a PR on the nf-core/sarek _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [x] Make sure your code lints (`nf-core lint .`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

**Comment by lassefolkersen** [comment]
Ok, I think this is ready for review now. Intended behaviour:
* if csv file has 0 or 1 lines. Full stop, no matter what. That can never be productive (and would skip subsequent checks anyway, because they get read as empty, #589 ).
* if csv file has >1 lines, check for presence of fields `patient` and `sample`, full stop if not present.
* remaining checks are unaltered - e.g. if `fastq_1` is present, and that file is not there it will check and stop (but that was already part of the code)

File: workflows/sarek.nf [hunk 1]
```diff
@@ -904,10 +904,28 @@ workflow.onComplete {
 */
 // Function to extract information (meta data + file(s)) from csv file(s)
 def extract_csv(csv_file) {
+
+    // check that the sample sheet is not 1 line or less, because it'll skip all subsequent checks if so.
+    new File(csv_file.toString()).withReader('UTF-8') { reader ->
+      def line, noOfLinesInSampleSheet = 0;
```
Comment (FriederikeHanssen) [inline workflows/sarek.nf:910]:
```suggestion
      def line, numberOfLinesInSampleSheet = 0;
```

**Review by FriederikeHanssen — APPROVED** [review]
Nice! Thanks for adding this. Just something small, I would propose renaming to "numberOfLinesInSampleSheet" or maybe at least `num` .

**Comment by FriederikeHanssen** [comment]
Do  you think it would be possible to add a check that each header field is one of (patient, sample, status, gender, lane, fastq1, fastq2, bam, bai, table, cram, crai, vcf)? (I think those are all but I would need to double check)

File: workflows/sarek.nf (hunk 1 above)
Comment (lassefolkersen) [reply workflows/sarek.nf]: ok good idea

**Comment by asp8200** [comment]
> Do you think it would be possible to add a check that each header field is one of (patient, sample, status, gender, lane, fastq1, fastq2, bam, bai, table, cram, crai, vcf)? (I think those are all but I would need to double check)

I was thinking along the same lines. Could we do something like

```
accepted_headers = [patient, sample, status, gender, lane, fastq1, fastq2, bam, bai, table, cram, crai, vcf]

for each header_field:
    if header_field not in accepted_headers then throw error
```

**Comment by lassefolkersen** [comment]
> > Do you think it would be possible to add a check that each header field is one of (patient, sample, status, gender, lane, fastq1, fastq2, bam, bai, table, cram, crai, vcf)? (I think those are all but I would need to double check)
> 
> I was thinking along the same lines. Could we do something like
> 
> ```
> accepted_headers = ['patient', 'sample', 'status', 'gender', 'lane', 'fastq1', 'fastq2', 'bam', 'bai', 'table', 'cram', 'crai', 'vcf']
> 
> for each header_field in headers:
>     if header_field not in accepted_headers then throw error
> ```

No, I don't like that solution. Because if you then do `-step variant_calling` (a bam/cram as input) or `-step annotation`(a vcf as input), then all those accepted_headers, except for `patient` and `sample` are _not_ required anymore. So it would effectively hinder the ability to enter Sarek at a later stage. 

So any solution has to be conditional on `-step`.

**Comment by FriederikeHanssen** [comment]
> No, I don't like that solution. Because if you then do `-step variant_calling` (a bam/cram as input) or `-step annotation`(a vcf as input), then all those accepted_headers, except for `patient` and `sample` are _not_ required anymore. So it would effectively hinder the ability to enter Sarek at a later stage.

I am not sure I follow. The way I understand Anders proposal, it would mean that a header could just not be `mysillyheaderfield` but we would at least check that it is in general part of correct list and then throw an error 
```
Unknown header field. Accepted columns are: 
```
. I agree it wouldn't check at all that the combination are proper, i.e. `--step annotation` should only have  `patient,sample,vcf` but would at least prevent typos and so on, but i don't get why it wouldn't advance

**Comment by FriederikeHanssen** [comment]
> > No, I don't like that solution. Because if you then do `-step variant_calling` (a bam/cram as input) or `-step annotation`(a vcf as input), then all those accepted_headers, except for `patient` and `sample` are _not_ required anymore. So it would effectively hinder the ability to enter Sarek at a later stage.
> 
> I am not sure I follow. The way I understand Anders proposal, it would mean that a header could just not be `mysillyheaderfield` but we would at least check that it is in general part of correct list and then throw an error
> 
> ```
> Unknown header field. Accepted columns are: 
> ```
> 
> . I agree it wouldn't check at all that the combination are proper, i.e. `--step annotation` should only have `patient,sample,vcf` but would at least prevent typos and so on, but i don't get why it wouldn't advance

But either way we can save this for the next PR :)

**Pull request merged**